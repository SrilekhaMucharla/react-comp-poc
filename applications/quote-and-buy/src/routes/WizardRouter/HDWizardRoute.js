import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Route, useLocation } from 'react-router-dom';
import _ from 'lodash';
import { updateEpticaId as updateEpticaIdAction } from '../../redux-thunk/actions';

const HDWizardRoute = ({
    handleForward,
    handleBackward,
    handleSkip,
    path,
    WizardPage,
    pageId,
    epticaId,
    updateEpticaId,
    pageMetadata,
    personalDetails
}) => {
    useEffect(() => {
        updateEpticaId(epticaId);
    }, [epticaId]);

    const isAdditionalDriver = !_.get(useLocation(), 'state.isPolicyHolder', true);
    const mappedPageMetadata = {
        ...pageMetadata,
        page_name: personalDetails ? `${pageMetadata.page_name}_${isAdditionalDriver ? 'AD' : 'MD'}` : pageMetadata.page_name
    };

    return (
        <Route
            exact
            path={path}
            render={
                (props) => (
                    <WizardPage
                        {...props}
                        handleForward={handleForward}
                        handleBackward={handleBackward}
                        handleSkip={handleSkip}
                        pageMetadata={mappedPageMetadata}
                        pageId={pageId} />
                )} />
    );
};

const mapDispatchToProps = {
    updateEpticaId: updateEpticaIdAction
};

HDWizardRoute.propTypes = {
    pageId: PropTypes.string.isRequired,
    handleBackward: PropTypes.func.isRequired,
    handleForward: PropTypes.func.isRequired,
    handleSkip: PropTypes.func.isRequired,
    path: PropTypes.string.isRequired,
    WizardPage: PropTypes.elementType.isRequired,
    epticaId: PropTypes.string.isRequired,
    updateEpticaId: PropTypes.func.isRequired,
    pageMetadata: PropTypes.shape({
        page_name: PropTypes.string.isRequired,
        page_type: PropTypes.string.isRequired,
        sales_journey_type: PropTypes.string.isRequired
    }).isRequired,
    personalDetails: PropTypes.bool.isRequired
};

export default connect(null, mapDispatchToProps)(HDWizardRoute);
