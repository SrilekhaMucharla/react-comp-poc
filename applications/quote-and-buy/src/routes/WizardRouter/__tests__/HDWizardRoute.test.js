import React from 'react';
import { mount, shallow } from 'enzyme';
import { Route, MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import HDWizardRoute from '../HDWizardRoute';

describe('HDWizardRoute', () => {
    // given
    const handleForward = jest.fn();
    const handleBackward = jest.fn();
    const path = 'test-path';
    const pageMetadata = {
        page_name: 'page-name'
    };
    // eslint-disable-next-line react/prop-types
    const WizardPage = ({ children }) => <div>{children}</div>;

    const initialState = {
        epticaId: 854
    };

    const middlewares = [];
    const mockStore = configureStore(middlewares);

    const store = mockStore(initialState);

    it('should render correctly and match the snapshot', () => {
        // given
        const wrapper = shallow(
            <Provider store={store}>
                <MemoryRouter initialEntries={[path]} initialIndex={0}>
                    <HDWizardRoute
                        handleBackward={handleBackward}
                        handleForward={handleForward}
                        path={path}
                        WizardPage={WizardPage}
                        pageMetadata={pageMetadata}
                        personalDetails />
                </MemoryRouter>
            </Provider>
        );

        // then
        expect(wrapper).toMatchSnapshot();
    });

    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter initialEntries={[path]} initialIndex={0}>
                <HDWizardRoute
                    handleBackward={handleBackward}
                    handleForward={handleForward}
                    path={path}
                    WizardPage={WizardPage}
                    pageMetadata={pageMetadata}
                    personalDetails />
            </MemoryRouter>
        </Provider>
    );

    it('should contain WizardPage', () => {
        // then
        expect(wrapper.find(WizardPage)).toHaveLength(1);
    });

    it('should pass path to the Route', () => {
        // when
        const route = wrapper.find(Route);
        // then
        expect(route.props().path).toBe(path);
    });

    it('should pass handleForward and handleBackward to the WizardPage', () => {
        // when
        const page = wrapper.find(WizardPage);
        // then
        expect(page.props().handleBackward).not.toBeUndefined();
        expect(page.props().handleForward).not.toBeUndefined();
    });
});
