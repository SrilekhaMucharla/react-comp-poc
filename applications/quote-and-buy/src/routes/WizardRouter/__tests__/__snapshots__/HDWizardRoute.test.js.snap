// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HDWizardRoute should render correctly and match the snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        "test-path",
      ]
    }
    initialIndex={0}
  >
    <Connect(HDWizardRoute)
      WizardPage={[Function]}
      handleBackward={[MockFunction]}
      handleForward={[MockFunction]}
      pageMetadata={
        Object {
          "page_name": "page-name",
        }
      }
      path="test-path"
      personalDetails={true}
    />
  </MemoryRouter>
</ContextProvider>
`;
